source/dofs/dof_handler_policy.cc

```diff
@@ -4140,46 +4773,36 @@ namespace internal
+
+        // --------- Phase 2: eliminate dof duplicates on all cells:
+        //                    - un-numerate dofs on interfaces to ghost cells
+        //                      that we don't own
+        //                    - in case of hp::DoFHandler, unify dofs
+        //                    - re-enumerate the remaining ones
         std::vector<dealii::types::global_dof_index> renumbering(
-          n_initial_local_dofs);
-        for (dealii::types::global_dof_index i = 0; i < renumbering.size(); ++i)
-          renumbering[i] = i;
-
-        {
-          std::vector<dealii::types::global_dof_index> local_dof_indices;
-
-          for (auto cell : dof_handler->active_cell_iterators())
-            if (cell->is_ghost() && (cell->subdomain_id() <
-                                     triangulation->locally_owned_subdomain()))
-              {
-                // we found a neighboring ghost cell whose subdomain
-                // is "stronger" than our own subdomain
-
-                // delete all dofs that live there and that we have
-                // previously assigned a number to (i.e. the ones on
-                // the interface)
-                local_dof_indices.resize(cell->get_fe().dofs_per_cell);
-                cell->get_dof_indices(local_dof_indices);
-                for (auto &local_dof_index : local_dof_indices)
-                  if (local_dof_index != numbers::invalid_dof_index)
-                    renumbering[local_dof_index] = numbers::invalid_dof_index;
-              }
-        }
-
-
-        // make the remaining indices consecutive
-        dealii::types::global_dof_index n_locally_owned_dofs = 0;
-        for (auto &new_index : renumbering)
-          if (new_index != numbers::invalid_dof_index)
-            new_index = n_locally_owned_dofs++;
+          n_initial_local_dofs, enumeration_dof_index);
+
+        // first, we invalidate degrees of freedom first that belong to
+        // processors of a lower rank, from which we will receive the
+        // final and lower degree of freedom later.
+        Implementation::invalidate_dof_indices_on_weaker_ghost_cells(
+          renumbering, subdomain_id, *dof_handler);
+
+        // then, we identify dof duplicates if a hp::DoFHandler is used,
+        // and enumerate all valid degrees of freedom consecutively.
+        // thus, we finally receive the correct number of locally owned
+        // cells after this step.
+        //
+        // the order in which we call these methods in phase 2 is important,
+        // since we want to clarify ownership of degrees of freedom before
+        // we actually unify their indices. otherwise, we could end up
+        // having a degee of freedom to which only invalid indices will
+        // be assigned.
+        const types::global_dof_index n_locally_owned_dofs =
+          Implementation::unify_and_enumerate_dof_indices(renumbering,
+                                                          *dof_handler);
```

Comment:
should this function really be called `unify_and_reenumerate_dof_indices`? you are not enumerating from scratch, after all